File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week2/day1/AcWing97/main.go
package main
import "fmt"

/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/28 14:49
 */
const mod = 9901

func qmi(a, b, p int) int {
    res := 1 % p
    for b > 0 {
        if b & 1 > 0 {res = res * a % p}
        a = a * a % p
        b >>= 1
    }
    return res
}

func sum(p, k int) int {
    if  k == 0 {return 1}
    if k % 2 == 0 {return (p % mod * sum(p, k - 1) + 1) % mod}
    return (1 + qmi(p, k / 2 + 1, mod)) * sum(p, k / 2) % mod
}

func main() {
    a, b := 0, 0
    fmt.Scan(&a, &b)
    res := 1
    for i := 2; i <= a; i++ {
        s := 0
        for a % i == 0 {
            s++
            a /= i
        }
        if s > 0 {res = res * sum(i, s * b) % mod}
    }
    if a == 0 {res = 0}
    fmt.Println(res)
}

File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week2/day1/AcWing97/main.py
#author: 袁佳林
def qmi(a, b):

    a %= mod
    res = 1
    while b:
        if b & 1:
            res = a * res % mod
        a = a * a % mod
        b >>= 1
    return res

def sum(p, k):
    if k == 0: return 1
    if k % 2 == 0:
        return (p % mod * sum(p, k - 1) + 1) % mod
    return (1 + qmi(p, k // 2 + 1)) * sum(p, k // 2) % mod

def main():
    a, b = map(int, input().split())
    res = 1
    i = 2
    while i <= a / i:
        s = 0
        while a % i == 0:
            s += 1
            a //= i
        if s:
            res = res * sum(i, s * b) % mod
        i += 1
    if a > 1:
        res = res * sum(a, b) % mod
    if not a:
        res = 0
    print(res)

if __name__ == "__main__":
    mod = 9901
    main()

